internal/terraform/rootmodule/root_module.go

```diff
@@ -378,6 +378,67 @@ func (rm *rootModule) ExecuteTerraformInit(ctx context.Context) error {
 	return rm.tfExec.Init(ctx)
 }
 
+func (rm *rootModule) ExecuteTerraformValidate(ctx context.Context) (map[string]hcl.Diagnostics, error) {
+	diagsMap := make(map[string]hcl.Diagnostics)
+
+	if !rm.IsTerraformAvailable() {
+		if err := rm.discoverTerraformExecutor(ctx); err != nil {
+			return diagsMap, err
+		}
+	}
+
+	if !rm.IsParsed() {
+		if err := rm.ParseFiles(); err != nil {
+			return diagsMap, err
+		}
+	}
+
+	// an entry for each file should exist, even if there are no diags
+	for filename := range rm.pFilesMap {
+		diagsMap[filename] = make(hcl.Diagnostics, 0)
+	}
+
+	validationDiags, err := rm.tfExec.Validate(ctx)
+	if err != nil {
+		return diagsMap, err
+	}
+
+	// tfjson.Diagnostic is a conversion of an internal diag to terraform core,
+	// tfdiags, which is effectively based on hcl.Diagnostic.
+	// This process is really just converting it back to hcl.Diagnotic
+	// since it is the defacto diagnostic type for our codebase currently
+	// https://github.com/hashicorp/terraform/blob/ae025248cc0712bf53c675dc2fe77af4276dd5cc/command/validate.go#L138
```

Comment:
It may be convertible now, but even Terraform itself has its own type for interpreting diags, so I wouldn't be surprised if at some point we find out that `hcl` types just aren't sufficient to interpret `validate` diags.

I think that returning `[]tfjson.Diagnostic` and then merging it into `[]lsp.Diagnostic` outside of this package only when we need to, may be safer approach.

Reply:
Could we punt on that for a future PR? It helped to only work in `hcl` until it gets to the diagnostics notifier package. But if you felt strongly I could create a second `Publish` method.

Reply:
Yeah, I suppose it's fine - I reckon we'll need to revisit this logic soon anyway to deliver on the promise of integrating tflint and somehow more "generalizing" the interface, or at least making it extendable.

So I am fairly confident this will need to change in the future, but I'm ok with leaving it for another PR. 👌